### PR TITLE
Implement ability to use two search approaches in knoema client

### PR DESCRIPTION
**Pull request by FluttercordKN**

### CODE REVIEW

File: Knoem.Client.UnitTests/KnoemaClientTests.cs [hunk 1]
```diff
@@ -3,7 +3,6 @@
 using System.Configuration;
 using System.IO;
 using System.Linq;
-
```
Comment (unicrus) [inline Knoem.Client.UnitTests/KnoemaClientTests.cs]: Revert the change.

File: Knoema.Client/Client.cs [hunk 2]
```diff
@@ -14,13 +14,15 @@
 using Knoema.Data;
 using Knoema.Meta;
 using Knoema.Search;
+using Knoema.Search.TimeseriesSearch;
 using Knoema.Upload;
 using Newtonsoft.Json;
```
Comment (unicrus) [inline Knoema.Client/Client.cs]: Organize usings.

File: Knoema.Client/Client.cs [hunk 3]
```diff
@@ -350,15 +398,15 @@ private static string GetAuthorizationHeaderValue(string clientId, string client
 					AuthProtoVersion);
 		}
 
-		private static Uri GetUri(string host, string token, string path, Dictionary<string, string> parameters = null)
+		private Uri GetUri(string host, string token, string path, Dictionary<string, string> parameters = null)
```
Comment (unicrus) [inline Knoema.Client/Client.cs]: Pass a scheme as a parameter.

File: Knoema.Client/Search/FieldCollection.cs [hunk 4]
```diff
@@ -0,0 +1,34 @@
+﻿using System.Collections.Generic;
+using System.Runtime.Serialization;
+using Newtonsoft.Json;
```
Comment (unicrus) [inline Knoema.Client/Search/FieldCollection.cs]: Organize usings.

File: Knoema.Client/Search/FieldCollection.cs [hunk 5]
```diff
@@ -0,0 +1,34 @@
...
+using Newtonsoft.Json.Linq;
+
+namespace Knoema
+{
+	public abstract class FieldCollection<T>
```
Comment (unicrus) [inline Knoema.Client/Search/FieldCollection.cs]: Add a description to the class.

File: Knoema.Client/Search/TimeseriesSearch/TimeseriesSearchFacets.cs [hunk 6]
```diff
@@ -0,0 +1,20 @@
+﻿using System.Collections.Generic;
+using System.Runtime.Serialization;
+using Newtonsoft.Json;
+using Newtonsoft.Json.Linq;
+using System.Linq;
```
Comment (unicrus) [inline Knoema.Client/Search/TimeseriesSearch/TimeseriesSearchFacets.cs]: Organize usings.

File: Knoema.Client/Client.cs (hunk 2 above)
Comment (FluttercordKN) [reply Knoema.Client/Client.cs]: done

File: Knoema.Client/Search/FieldCollection.cs (hunk 4 above)
Comment (FluttercordKN) [reply Knoema.Client/Search/FieldCollection.cs]: done

File: Knoema.Client/Search/FieldCollection.cs (hunk 5 above)
Comment (FluttercordKN) [reply Knoema.Client/Search/FieldCollection.cs]: done

File: Knoem.Client.UnitTests/KnoemaClientTests.cs (hunk 1 above)
Comment (FluttercordKN) [reply Knoem.Client.UnitTests/KnoemaClientTests.cs]: done

File: Knoema.Client/Client.cs (hunk 3 above)
Comment (FluttercordKN) [reply Knoema.Client/Client.cs]: done

File: Knoema.Client/Search/TimeseriesSearch/TimeseriesSearchFacets.cs (hunk 6 above)
Comment (FluttercordKN) [reply Knoema.Client/Search/TimeseriesSearch/TimeseriesSearchFacets.cs]: done

File: Knoema.Client/Client.cs [hunk 7]
```diff
@@ -36,22 +38,29 @@ public class Client
 
 		public int HttpTimeout { get; set; }
 
-		private Client()
+		private Client(bool ignoreCertErrors, string scheme)
```
Comment (PavelPimenov) [inline Knoema.Client/Client.cs]: Why do you need IgnoreCertErrors?
Why do you need scheme?

File: Knoema.Client/Client.cs [hunk 8]
```diff
@@ -63,8 +72,8 @@ public Client(string host, string token)
 			_token = token;
 		}
 
-		public Client(string host, string clientId, string clientSecret)
-			: this()
+		public Client(string host, string clientId, string clientSecret, bool ignoreCertErrors = false, string scheme = "")
```
Comment (PavelPimenov) [inline Knoema.Client/Client.cs]: Why cannot you just add one ctor with Ignorecert and scheme instead of changing all of the ctors. and still question above is valid

File: Knoema.Client/Client.cs (hunk 7 above)
Comment (FluttercordKN) [reply Knoema.Client/Client.cs]: these options allows to use API via HTTPS with invalid certificate

File: Knoema.Client/Client.cs (hunk 8 above)
Comment (FluttercordKN) [reply Knoema.Client/Client.cs]: How do you pass host, appId, appSecrec and token parameters then?

File: Knoema.Client/Client.cs (hunk 7 above)
Comment (PavelPimenov) [reply Knoema.Client/Client.cs]: I understand what it allows to do. but why do you need this? What is the reason? Because it looks like you are trying to workaround the server issue here

File: Knoema.Client/Client.cs (hunk 8 above)
Comment (PavelPimenov) [reply Knoema.Client/Client.cs]: I mean one more ctor like this Client(host, clientId, clientSecret, ignoreCertErrors, scheme)

File: Knoema.Client/Search/TimeseriesSearch/TimeseriesSearchFacets.cs [hunk 9]
```diff
@@ -0,0 +1,20 @@
...
+{
+	public class TimeseriesSearchFacets
+	{
+		public TimeseriesSearchDataSourceInfos DataSourceInfos { get; set; }
+		public Dictionary<string, int> DataSources { get; set; }
+		public int Forecasts { get; set; }
+		public Dictionary<string, int> Frequencies { get; set; }
+		public Dictionary<string, int> LastUpdates { get; set; }
+		public int LongTimeseries { get; set; }
+		public TimeseriesSearchRegionLinkInfos RegionLinkInfos { get; set; }
```
Comment (BillSytin) [inline Knoema.Client/Search/TimeseriesSearch/TimeseriesSearchFacets.cs]: special class is redundant, it can be just a Dictionary<string, GeographyRegionViewModel>

File: Knoema.Client/Search/TimeseriesSearch/TimeseriesSearchFacets.cs [hunk 10]
```diff
@@ -0,0 +1,20 @@
...
+using System.Runtime.Serialization;
+using Newtonsoft.Json;
+using Newtonsoft.Json.Linq;
+using System.Linq;
+
+namespace Knoema.Search.TimeseriesSearch
+{
+	public class TimeseriesSearchFacets
+	{
+		public TimeseriesSearchDataSourceInfos DataSourceInfos { get; set; }
```
Comment (BillSytin) [inline Knoema.Client/Search/TimeseriesSearch/TimeseriesSearchFacets.cs]: special class is redundant, it can be just Dictionary<string, DatasetSourceItemViewModel>

File: Knoema.Client/Search/TimeseriesSearch/TimeseriesSearchLocation.cs [hunk 11]
```diff
@@ -0,0 +1,13 @@
...
+using System.Collections.Generic;
+
+namespace Knoema.Search.TimeseriesSearch
+{
+	public class TimeseriesSearchLocation
+	{
+		public int DatasetKey { get; set; }
+		public DateTime Expires { get; set; }
+		public string Key { get; set; }
+		public IEnumerable<TimeseriesSearchPositionItem> Position { get; set; }
```
Comment (BillSytin) [inline Knoema.Client/Search/TimeseriesSearch/TimeseriesSearchLocation.cs]: No need to make it typed, it can be just an 'Object'.
Then server content of this field could change with no need to update clients.

File: Knoema.Client/Search/TimeseriesSearch/TimeseriesSearchFacets.cs (hunk 9 above)
Comment (FluttercordKN) [reply Knoema.Client/Search/TimeseriesSearch/TimeseriesSearchFacets.cs]: done

File: Knoema.Client/Search/TimeseriesSearch/TimeseriesSearchFacets.cs (hunk 10 above)
Comment (FluttercordKN) [reply Knoema.Client/Search/TimeseriesSearch/TimeseriesSearchFacets.cs]: done

File: Knoema.Client/Search/TimeseriesSearch/TimeseriesSearchLocation.cs (hunk 11 above)
Comment (FluttercordKN) [reply Knoema.Client/Search/TimeseriesSearch/TimeseriesSearchLocation.cs]: done

File: Knoema.Client/Search/TimeseriesSearch/TimeseriesSearchRequestFacets.cs [hunk 12]
```diff
@@ -0,0 +1,14 @@
+﻿using System.Collections.Generic;
+
+namespace Knoema.Search.TimeseriesSearch
+{
+	public class TimeseriesSearchRequestFacets
```
[the remark this answers is not in the document]
Comment (FluttercordKN) [reply Knoema.Client/Search/TimeseriesSearch/TimeseriesSearchRequestFacets.cs]: done

File: Knoema.Client/Client.cs [hunk 13]
```diff
@@ -36,22 +38,29 @@ public class Client
...
-		private Client()
+		private Client(bool ignoreCertErrors, string scheme)
 		{
 			HttpTimeout = DefaultHttpTimeout;
+
+			_scheme = scheme;
+			if (string.IsNullOrEmpty(_scheme))
+				_scheme = Uri.UriSchemeHttp;
+
+			if (ignoreCertErrors)
```
Comment (BillSytin) [inline Knoema.Client/Client.cs]: parameter 'ignoreCertErrors' has wrong meaning,
it will disable certificate errors for all connections, not only for this client

Comment (FluttercordKN) [reply Knoema.Client/Client.cs]: done

File: Knoema.Client/Search/FieldCollection.cs [hunk 14]
```diff
@@ -0,0 +1,39 @@
...
+					var item = pair.Value.ToObject<T>();
+					AttachKeyToObject(pair.Key, item);
+					items.Add(item);
+				}
+				Items = items;
+				_fields = null;
+			}
+		}
+
+		protected abstract void AttachKeyToObject(string key, T obj);
```
[the remark this answers is not in the document]
Comment (AlexPhilippov) [reply Knoema.Client/Search/FieldCollection.cs]: So?

File: Knoema.Client/Search/TimeseriesSearch/GeographyRegion.cs [hunk 15]
```diff
@@ -0,0 +1,12 @@
...
+namespace Knoema.Search.TimeseriesSearch
+{
+	public class GeographyRegion
+	{
+		public string Id { get; set; }
+		public string LocalizedName { get; set; }
+		public string Name { get; set; }
+		public IEnumerable<string> Parents { get; set; }
+	}
+}
```
Comment (AlexPhilippov) [inline Knoema.Client/Search/TimeseriesSearch/GeographyRegion.cs]: Also, we usually end files with an empty line. You can see git is also showing warning there.

File: Knoema.Client/Search/FieldCollection.cs (hunk 14 above)
[the remark this answers is not in the document]
Comment (FluttercordKN) [reply Knoema.Client/Search/FieldCollection.cs]: class removed

File: Knoema.Client/Search/TimeseriesSearch/GeographyRegion.cs (hunk 15 above)
Comment (FluttercordKN) [reply Knoema.Client/Search/TimeseriesSearch/GeographyRegion.cs]: done

**Comment by AlexPhilippov** [comment]
Moving to https://github.com/Knoema/knoema-csharp-driver/pull/44
